src/sources/stdin.rs

```diff
@@ -151,7 +161,7 @@ mod tests {
         let buf = Cursor::new(String::from("hello world\nhello world again"));
 
         let mut rt = Runtime::new().unwrap();
-        let source = stdin_source(buf, config, tx);
+        let source = stdin_source(buf, config, ShutdownSignal::noop(), tx);
```

Comment:
For this test, a proper shutdown isn't needed since the source will shutdown on it's own once it reads through the whole of 
```
Cursor::new(String::from("hello world\nhello world again"));
```

Reply:
If we reload this source and there is no eof to the stdin, we can't shut it down? (which I think is correct) but this is an odd behavior. So if we end up say restarting an app multiple times for stdin, the following sources won't be able to acquire the lock. This seems like an issue.

Reply:
Yes, this is a variation of the issue I mentioned in the PR opening comment, and since this was an issue even before this PR I decided to open a separate issue for it, but it would be also fine in addressing this here if you for it. In that case comment on the open question up there.

Reply:
@ktff feel free to fix it here and close that issue as well.

Reply:
So using the builtin tokio stdin won't work. We moved away from that because it was causing issues with shutdown. Basically, its the same problem we have here.

What we really should do is write this in terms of a single static stdin thread that can use the `tokio::sync::broadcast` channel. The broadcast channel works well here because the receivers are clonable and therefore we can wire it up to many sources without needing to pass a new sender end.

So this then means that we can apply the take_until to the receiver end of the broadcast channel and just leave the bg stdin thread running. We can then be kinda smart here and have some sort of atomic bool that can be checked everytime the read(2) syscall ends and if the bool is true we exit the thread. This is really the only way to properly handle shutdown with stdin.

Reply:
Alright, let's do it that way then.